runtime/queries/metricsview_comparision_toplist.go

```diff
@@ -0,0 +1,413 @@
+
+			measureValues = append(measureValues, &runtimev1.MetricsViewComparisonValue{
+				MeasureName: name,
+				BaseValue:   v,
+			})
+		}
+
+		dv, err := pbutil.ToValue(values[0])
+		if err != nil {
+			return err
+		}
+
+		data = append(data, &runtimev1.MetricsViewComparisonRow{
+			DimensionValue: dv,
+			MeasureValues:  measureValues,
+		})
+	}
+
+	q.Result = &runtimev1.MetricsViewComparisonToplistResponse{
+		Rows: data,
+	}
+
+	return nil
+}
+
+func (q *MetricsViewComparisonToplist) executeComparisonToplist(ctx context.Context, olap drivers.OLAPStore, mv *runtimev1.MetricsView, priority int) error {
+	sql, args, err := q.buildMetricsComparisonTopListSQL(mv, olap.Dialect())
+	if err != nil {
+		return fmt.Errorf("error building query: %w", err)
+	}
+
+	rows, err := olap.Execute(ctx, &drivers.Statement{
+		Query:    sql,
+		Args:     args,
+		Priority: priority,
+	})
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	var data []*runtimev1.MetricsViewComparisonRow
+	for rows.Next() {
+		values, err := rows.SliceScan()
+		if err != nil {
+			return err
+		}
+		measureValues := []*runtimev1.MetricsViewComparisonValue{}
+
+		for i, name := range q.MeasureNames {
+			bv, err := pbutil.ToValue(values[1+i*4])
+			if err != nil {
+				return err
+			}
+
+			cv, err := pbutil.ToValue(values[2+i*4])
+			if err != nil {
+				return err
+			}
+
+			da, err := pbutil.ToValue(values[3+i*4])
+			if err != nil {
+				return err
+			}
+
+			dr, err := pbutil.ToValue(values[4+i*4])
+			if err != nil {
+				return err
+			}
+
+			measureValues = append(measureValues, &runtimev1.MetricsViewComparisonValue{
+				MeasureName:     name,
+				BaseValue:       bv,
+				ComparisonValue: cv,
+				DeltaAbs:        da,
+				DeltaRel:        dr,
+			})
+		}
+
+		dv, err := pbutil.ToValue(values[0])
+		if err != nil {
+			return err
+		}
+
+		data = append(data, &runtimev1.MetricsViewComparisonRow{
+			DimensionValue: dv,
+			MeasureValues:  measureValues,
+		})
+	}
+
+	q.Result = &runtimev1.MetricsViewComparisonToplistResponse{
+		Rows: data,
+	}
+
+	return nil
+}
+
+func timeRangeClause(timeRange *runtimev1.TimeRange, td string, args *[]any) string {
+	var clause string
+	if timeRange == nil {
+		return clause
+	}
+
+	if timeRange.Start != nil {
+		clause += fmt.Sprintf(" AND %s >= ?", td)
+		*args = append(*args, timeRange.Start.AsTime())
+	}
+
+	if timeRange.End != nil {
+		clause += fmt.Sprintf(" AND %s < ?", td)
+		*args = append(*args, timeRange.End.AsTime())
+	}
+
+	return clause
+}
+
+func (q *MetricsViewComparisonToplist) buildMetricsTopListSQL(mv *runtimev1.MetricsView, dialect drivers.Dialect) (string, []any, error) {
+	ms, err := resolveMeasures(mv, q.InlineMeasures, q.MeasureNames)
+	if err != nil {
+		return "", nil, err
+	}
+
+	dimName := safeName(q.DimensionName)
+	selectCols := []string{dimName}
+	measureMap := make(map[string]int)
+	for i, m := range ms {
+		measureMap[m.Name] = i
+		expr := fmt.Sprintf(`%s as "%s"`, m.Expression, m.Name)
```

Comment:
Should we use `safeName` for `m.Name` instead?

Reply:
No need, `m.Name` is generated automatically (not a user input).

Reply:
No need, `m.Name` is generated automatically (not a user input).